lib/semantics/check-select-stmt.cpp

```diff
@@ -0,0 +1,316 @@
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "check-select-stmt.h"
+#include "flang/parser/message.h"
+#include "flang/semantics/tools.h"
+
+namespace Fortran::semantics {
+
+class SelectCaseHelper {
+
+public:
+  SelectCaseHelper(parser::Messages &messages) : messages_{messages} {}
+
+  void ReportOverlapCaseValError(
+      parser::CharBlock &source1, parser::CharBlock &source2) {
+    // C1149
+    messages_
+        .Say(source2,
+            "SELECT CASE statement value at this location overlaps with the below location"_err_en_US)
+        .Attach(source1,
+            "SELECT CASE statement value overlaps with the above location"_err_en_US);
+  }
+
+  void Insert(std::int64_t intVal, parser::CharBlock source) {
+    auto selectCaseIntRangeValsIter =
+        std::find_if(selectCaseIntRangeVals_.begin(),
+            selectCaseIntRangeVals_.end(), [&intVal](const auto &mem) {
+              return ((!mem.fromVal_ || (intVal >= mem.fromVal_.value())) &&
+                  (!mem.toVal_ || (intVal <= mem.toVal_.value())));
+            });
+    if (selectCaseIntRangeValsIter != selectCaseIntRangeVals_.end()) {
+      ReportOverlapCaseValError(selectCaseIntRangeValsIter->source_, source);
+    }
+    auto selectCaseIntValsInsertStatus =
+        selectCaseIntVals_.insert(std::make_pair(intVal, source));
+    if (selectCaseIntValsInsertStatus.second == false) {
+      ReportOverlapCaseValError(
+          selectCaseIntValsInsertStatus.first->second, source);
+    }
+  }
+  void Insert(std::string strVal, parser::CharBlock source) {
+    auto selectCaseStringRangeValsIter = std::find_if(
+        selectCaseStringRangeVals_.begin(), selectCaseStringRangeVals_.end(),
+        [&strVal](const auto &mem) {
+          return (
+              (!mem.fromVal_ || (strVal.compare(mem.fromVal_.value()) >= 0)) &&
+              (!mem.toVal_ || (strVal.compare(mem.toVal_.value()) <= 0)));
+        });
+    if (selectCaseStringRangeValsIter != selectCaseStringRangeVals_.end()) {
+      ReportOverlapCaseValError(selectCaseStringRangeValsIter->source_, source);
+    }
+    auto selectCaseStringValsInsertStatus =
+        selectCaseStringVals_.insert(std::make_pair(strVal, source));
+    if (selectCaseStringValsInsertStatus.second == false) {
+      ReportOverlapCaseValError(
+          selectCaseStringValsInsertStatus.first->second, source);
+    }
+  }
+  void Insert(bool logicalVal, parser::CharBlock source) {
+    auto selectCaseLogicalValsInsertStatus =
+        selectCaseLogicalVals_.insert(std::make_pair(logicalVal, source));
+    if (selectCaseLogicalValsInsertStatus.second == false) {
+      ReportOverlapCaseValError(
+          selectCaseLogicalValsInsertStatus.first->second, source);
+    }
+  }
+  void Insert(std::optional<std::int64_t> lowerVal,
```

Comment:
Abstract these into a template rather than coping and pasting the code, please.

Reply:
I have abstracted common code into function templates.